it/src/main/java/com/google/cloud/teleport/it/spanner/DefaultSpannerResourceManager.java

```diff
@@ -0,0 +1,243 @@
+
+  private boolean hasInstance = false;
+  private boolean hasDatabase = false;
+
+  private final String projectId;
+  private final String instanceId;
+  private final String databaseId;
+  private final String region;
+
+  private final Spanner spanner;
+  private final InstanceAdminClient instanceAdminClient;
+  private final DatabaseAdminClient databaseAdminClient;
+
+  private DefaultSpannerResourceManager(Builder builder) {
+    this(
+        SpannerOptions.newBuilder().setProjectId(builder.projectId).build().getService(),
+        builder.testId,
+        builder.projectId,
+        builder.region);
+  }
+
+  @VisibleForTesting
+  DefaultSpannerResourceManager(Spanner spanner, String testId, String projectId, String region) {
+    if (testId.length() > MAX_BASE_ID_LENGTH) {
+      testId = generateNewId(testId, MAX_BASE_ID_LENGTH);
+    }
+    this.projectId = projectId;
+    this.instanceId = generateInstanceId(testId);
+    this.databaseId = generateDatabaseId(testId);
+
+    this.region = region;
+    this.spanner = spanner;
+    this.instanceAdminClient = spanner.getInstanceAdminClient();
+    this.databaseAdminClient = spanner.getDatabaseAdminClient();
+  }
+
+  public static Builder builder(String testId, String projectId, String region) {
+    return new Builder(testId, projectId, region);
+  }
+
+  private synchronized void createInstance() {
+    LOG.info("Creating instance {} in project {}.", instanceId, projectId);
+    InstanceInfo instanceInfo =
+        InstanceInfo.newBuilder(InstanceId.of(projectId, instanceId))
+            .setInstanceConfigId(InstanceConfigId.of(projectId, region))
+            .setDisplayName(instanceId)
+            .setNodeCount(1)
+            .build();
+    try {
+      instanceAdminClient.createInstance(instanceInfo).get();
+      hasInstance = true;
+      LOG.info("Successfully created instance {}.", instanceId);
+    } catch (ExecutionException | InterruptedException | SpannerException e) {
+      throw new SpannerResourceManagerException("Failed to create instance.", e);
+    }
+  }
+
+  private synchronized void createDatabase() {
+    LOG.info("Creating database {} in instance {}.", databaseId, instanceId);
+    try {
+      databaseAdminClient.createDatabase(instanceId, databaseId, ImmutableList.of()).get();
+      hasDatabase = true;
+      LOG.info("Successfully created database {}.", databaseId);
+    } catch (ExecutionException | InterruptedException | SpannerException e) {
+      throw new SpannerResourceManagerException("Failed to create database.", e);
+    }
+  }
+
+  private void checkIsUsable() throws IllegalStateException {
+    if (spanner.isClosed()) {
+      throw new IllegalStateException("Manager has cleaned up all resources and is unusable.");
+    }
+  }
+
+  @Override
+  public synchronized void createTable(String statement) throws IllegalStateException {
+    checkIsUsable();
+    if (!hasInstance) {
+      createInstance();
+    }
```

Comment:
The functions are separated because the database creation may fail while the instance creation succeeds. In that case, the manager will have an instance but no database.

Reply:
True, but the resource manager is unusable if any of these fail. If it has an instance but no database, and I call any other method, it won't work. So I think the approach should be "all or nothing". Create the instance AND the database. If both succeed, the resource manager is ready to handle other requests. If any of these 2 fail, throw an exception, cleanup the resources, etc. We can't run the integration test anyway. Right?

I'm OK with keeping this as separate methods, though I would still at least move the if check inside the method. Basically any new method in this resource manager that reads from or writes to the database will have to have this code block copy-pasted. So the smaller it is the better.